### DISCUSSION TITLE
get []  sometimes

### DISCUSSION THREAD

**Discussion by vaynevayne** [pmndrs/zustand#1885] [Bug report] [answered]

**Comment by dai-shi** [comment]
Can you create a reproduction with https://codesandbox.io/s/react-new ?

**Comment by vaynevayne** [comment]
```jsx

const  init = {
sortOrder:['1','b']
}
const appStoreSortOrder = useAppStore(state => state.sortOrder )
console.log('render', appStoreSortOrder) // ['1','b']

const fn =()=>{
 const sortList = appStoreSortOrder

                        while (sortList.length) {
                          const shift = sortList.shift()
                        
                        }

}


    
```

When fn is executed, the appStore becomes [] 

Is this normal behavior?

Is there any way to prevent unexpected changes?

For example, zusTand's readonly STate

**Comment by vaynevayne** [comment]
@dai-shi  What's the situation now?

**Reply by dai-shi** [reply]
I'm not sure if I understand the problem. I wonder if someone can help it.

**Reply by dbritto-dev** [reply]
@vaynevayne if you need a readonly state then you don't need a zustand store or any store at all; zustand stores are readable and writable.

I would use an obj (constant) + `Object.freeze()`

**Reply by vaynevayne** [reply]
Well, actually, all the states you get from useStore are read-only, right?

So is it reasonable for zustand to wrap Object.freeze() by default? Are there any other problems?

@dbritto-dev

**Reply by dbritto-dev** [reply, marked as answer]
> Well, actually, all the states you get from useStore are read-only, right?
> 
> So is it reasonable for zustand to wrap Object.freeze() by default? Are there any other problems?
> 
> @dbritto-dev 

Actually, only primitives are readonly (you can't mutate them, just reassign them) in JS. What you are doing is mutating an object directly (which is wrong in a react context). BTW, what you get from `useStore` is not the store itself if just the result of a given selector, even if the selector if the "store" itself. Nothing avoids mutate that result (object) directly but is wrong in a react context due to that mutation only exists on the scope component so if you read the store on another component the values still the same. 

Mutate an object direcly, you can do it but you should not do it at least in a react context. You must read this article: https://react.dev/learn/updating-objects-in-state
